threshold_encryption/ThresholdEncryption.cpp

```diff
@@ -0,0 +1,147 @@
+#include "TEBase.h"
+#include "TEDecryptSet.h"
+#include <tools/utils.h>
+#include <valarray>
+
+libBLS::CipherResult ThresholdEncryption::encrypt(
+    const std::string& message, const TEPublicKey& commonPublic ) {
+    TEBase::initializeIfNecessary();
+
+    libBLS::CipherResult cypher =
+        libBLS::TE::encryptWithAES( message, commonPublic.getPublicKeyRaw() );
+
+    libBLS::TE::checkCypher( cypher.ciphertext->key );
+
+    return cypher;
+}
+
+bool ThresholdEncryption::validateEncryption( const libBLS::CipheredKey& ciphertext ) {
+    TEBase::initializeIfNecessary();
+
+    auto [U, V, W] = ciphertext;
+
+    libff::alt_bn128_G1 H = libBLS::TE::HashToGroup( U, V );
+
+    libff::alt_bn128_GT fst, snd;
+
+    // pairing( W, P ) == pairing( H, U )
+    fst = libff::alt_bn128_ate_reduced_pairing( W, libff::alt_bn128_G2::one() );
+    snd = libff::alt_bn128_ate_reduced_pairing( H, U );
+
+    return fst == snd;
+}
+
+TEDecryptionShare ThresholdEncryption::partialDecrypt(
+    const libBLS::CipheredKey& ciphertext, const TEPrivateKeyShare& pkeyShare ) {
+    TEBase::initializeIfNecessary();
+
+    libBLS::TE::checkCypher( ciphertext );
+
+    libff::alt_bn128_G2 decryption_share =
+        libBLS::TE::getDecryptionShare( ciphertext, pkeyShare.getPrivateKeyRaw() );
+
+    if ( decryption_share.is_zero() || !decryption_share.is_well_formed() ) {
+        throw libBLS::ThresholdUtils::IsNotWellFormed( "zero decrypt" );
+    }
+
+    TEDecryptionShare share( pkeyShare.getSignerIndex(), decryption_share );
+
+    return share;
+}
+
+bool ThresholdEncryption::validateDecryptionShare( const libBLS::CipheredKey& cipherText,
+    const TEDecryptionShare& decryptionShare, const TEPublicKeyShare& publicKey ) {
+    TEBase::initializeIfNecessary();
+
+    libBLS::TE::checkCypher( cipherText );
+
+    if ( !decryptionShare.validate() ) {
+        throw libBLS::ThresholdUtils::IsNotWellFormed( "Invalid decryption share" );
+    }
+
+    return libBLS::TE::Verify(
+        cipherText, decryptionShare.getShareRaw(), publicKey.getPublicKeyRaw() );
+}
+
+std::string ThresholdEncryption::combineShares(
+    const libBLS::Ciphertext& cyphertext, TEDecryptSet& decryptionSet ) {
+    TEBase::initializeIfNecessary();
+
+    libBLS::TE::checkCypher( cyphertext.key );
+
+    if ( !decryptionSet.canMerge() ) {
+        auto status = decryptionSet.getMergeStatus();
+        if ( status == TEDecryptSet::MergeStatus::NOT_ENOUGH_SHARES ) {
+            throw libBLS::ThresholdUtils::IsNotWellFormed(
+                "Not enough elements to decrypt message" );
+        } else if ( status == TEDecryptSet::MergeStatus::ALREADY_MERGED ) {
+            throw libBLS::ThresholdUtils::IsNotWellFormed( "Already merged" );
+        }
+    }
+
+    libBLS::TE te( decryptionSet );
+    auto aesKey = te.CombineShares( cyphertext.key, decryptionSet.getSharesRaw() );
+    auto decriptedMessage = libBLS::ThresholdUtils::aesDecrypt( *cyphertext.data, aesKey );
```

Comment:
Please cyphertext.data for null before dereferencing with *. We have it in code guidelines (each time we use a pointer that may be null, we need to have a null check). This is done to prevent segmentation faults.